pvlib/pvsystem.py

```diff
@@ -1929,23 +1938,253 @@ def i_from_v(resistance_shunt, resistance_series, nNsVth, voltage,
 
     # asarray turns Series into arrays so that we don't have to worry
     # about multidimensional broadcasting failing
-    Rsh = np.asarray(resistance_shunt)
-    Rs = np.asarray(resistance_series)
-    I0 = np.asarray(saturation_current)
-    IL = np.asarray(photocurrent)
-    V = np.asarray(voltage)
-
-    argW = (Rs*I0*Rsh *
-            np.exp(Rsh*(Rs*(IL+I0)+V) / (nNsVth*(Rs+Rsh))) /
-            (nNsVth*(Rs + Rsh)))
+    Rsh = np.asarray(resistance_shunt, np.float64)
+    Rs = np.asarray(resistance_series, np.float64)
+    nNsVth = np.asarray(nNsVth, np.float64)
+    V = np.asarray(voltage, np.float64)
+    I0 = np.asarray(saturation_current, np.float64)
+    IL = np.asarray(photocurrent, np.float64)
+
+    # This transforms any ideal Rsh=np.inf into Gsh=0., which is generally more numerically stable
+    Gsh = 1./Rsh
+
+    # argW cannot be float128
+    argW = \
+Rs*I0/(nNsVth*(Rs*Gsh + 1.))*np.exp((Rs*(IL + I0) + V)/(nNsVth*(Rs*Gsh + 1.)))
     lambertwterm = lambertw(argW).real
 
     # Eqn. 4 in Jain and Kapoor, 2004
-    I = -V/(Rs + Rsh) - (nNsVth/Rs)*lambertwterm + Rsh*(IL + I0)/(Rs + Rsh)
+    I = (IL + I0 - V*Gsh)/(Rs*Gsh + 1.) - (nNsVth/Rs)*lambertwterm
```

Comment:
Let's leave the original in as a comment to accompany the reference, it helps if someone wants to understand the source of the equation.